src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,200 @@
+        }
+
+        if (collection.isEmpty()) {
+            throw new RuntimeException("Can't add elements from empty collection " + collection);
+        }
+        Object[] objects = collection.toArray();
+
+        if (index == size) {
+            for (Object o : objects) {
+                add((T) o);
+            }
+        } else {
+            for (Object o : objects) {
+                linkBefore((T) o, getNode(index));
+                index++;
+            }
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> setNode = getNode(index);
+        final T oldValue = setNode.item;
+        setNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> current = head; current != null; current = current.next) {
+                if (current.item == null) {
+                    unlink(current);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> current = head; current != null; current = current.next) {
+                if (object.equals(current.item)) {
+                    unlink(current);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkLast(T value) {
+        final Node<T> l = tail;
+        final Node<T> newNode = new Node<>(l, value, null);
+        tail = newNode;
+
+        if (l == null) {
+            head = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> succeeding) {
+        if (succeeding == null) {
+            throw new IndexOutOfBoundsException("Can't link element "
+                    + value + " before null Node " + succeeding);
+        }
+
+        final Node<T> precedent = succeeding.prev;
+        final Node<T> newNode = new Node<>(precedent, value, succeeding);
+        succeeding.prev = newNode;
+
+        if (precedent == null) {
+            head = newNode;
+        } else {
+            precedent.next = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> result = head;
+        checkElementIndex(index);
+
+        for (int i = 0; i < index; i++) {
+            result = result.next;
+        }
```

Comment:
It should be 2 cases of iteration:
* form the `head` to `tail` if the `index` is closer to the `head`
* form the `tail` to `head` if the `index` is closer to the `tail`

Reply:
> It should be 2 cases of iteration:
> 
> * form the `head` to `tail` if the `index` is closer to the `head`
> * form the `tail` to `head` if the `index` is closer to the `tail`

done